lib/realtime/tenants/connect.ex

```diff
@@ -0,0 +1,90 @@
+defmodule Realtime.Tenants.Connect do
+  @moduledoc """
+  This module is responsible for attempting to connect to a tenant's database and store the DBConnection in a Syn registry.
+  """
+  use GenServer
+
+  require Logger
+
+  alias Realtime.Helpers
+  alias Realtime.Tenants
+
+  defstruct tenant_id: nil, db_conn_reference: nil
+
+  @doc """
+  Returns the database connection for a tenant. If the tenant is not connected, it will attempt to connect to the tenant's database.
+  """
+  @spec lookup_or_start_connection(binary()) :: {:ok, DBConnection.t()} | {:error, term()}
+  def lookup_or_start_connection(tenant_id) do
+    case get_status(tenant_id) do
+      :undefined -> call_external_node(tenant_id)
+      {:ok, conn} -> {:ok, conn}
+      _ -> {:error, :tenant_database_unavailable}
+    end
+  end
+
+  @doc """
+  Connects to a tenant's database and stores the DBConnection in the process :syn metadata
+  """
+  @spec connect(binary()) :: {:ok, DBConnection.t()} | {:error, term()}
+  def connect(tenant_id) do
+    supervisor = {:via, PartitionSupervisor, {Realtime.Tenants.Connect.DynamicSupervisor, self()}}
+    spec = {__MODULE__, tenant_id: tenant_id}
+
+    case DynamicSupervisor.start_child(supervisor, spec) do
+      {:ok, _} -> get_status(tenant_id)
+      {:error, {:already_started, _}} -> get_status(tenant_id)
+      _ -> {:error, :tenant_database_unavailable}
+    end
+  end
+
+  def start_link(tenant_id: tenant_id) do
+    name = {__MODULE__, tenant_id, %{conn: nil}}
+    GenServer.start_link(__MODULE__, %__MODULE__{tenant_id: tenant_id}, name: {:via, :syn, name})
+  end
+
+  ## GenServer callbacks
+  # Needs to be done on init/1 to guarantee the GenServer only starts if we are able to connect to the database
+  @impl GenServer
+  def init(%{tenant_id: tenant_id} = state) do
+    with tenant when not is_nil(tenant) <- Tenants.Cache.get_tenant_by_external_id(tenant_id),
+         res <- Helpers.check_tenant_connection(tenant) do
+      case res do
+        {:ok, conn} ->
+          :syn.update_registry(__MODULE__, tenant_id, fn _pid, meta -> %{meta | conn: conn} end)
+          state = %{state | db_conn_reference: Process.monitor(conn)}
+
+          {:ok, state}
+
+        {:error, error} ->
+          Logger.error("Error connecting to tenant database: #{inspect(error)}")
+          {:stop, :normal, state}
```

Comment:
I don't believe this is a valid return type. You can use `{:stop, :normal}` or `:ignore`. Although you should check what happens based on your supervisor and restart strategies. I would also manually double-check your choice of supervisor strategy + genserver restart is fine with this return type (some options means it could potentially retry or keep the child specification in the supervisor).

Reply:
this was a wrong copy paste I believe 😓 we just need `{:stop, :normal}` in this scenario so will use that 👍